File: Website/App_Code/abstract/pages/CMasterPage.vb
Imports Microsoft.VisualBasic

Public MustInherit Class CMasterPage : Inherits MasterPage

    Private Sub Page_PreRender(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.PreRender
        If String.IsNullOrEmpty(PageHeading) Then PageHeading = Page.Title
        If String.IsNullOrEmpty(MenuTitle) Then MenuTitle = CConfig.MenuTitle
        If Main.Controls.Count = 0 AndAlso Not IsNothing(Fieldset) Then Fieldset.Visible = False
        If Not Page.IsPostBack AndAlso CSession.IsLoggedIn Then SchemaMembership.CClick.Log()
    End Sub

    Public MustOverride ReadOnly Property Main() As ContentPlaceHolder
    Public MustOverride ReadOnly Property Fieldset() As HtmlControls.HtmlGenericControl

    Public MustOverride Property PageHeading() As String
    Public MustOverride Sub AddMenuLeft(ByVal name As String, ByVal url As String, ByVal selected As Boolean, ByVal tooltip As String)
    Public MustOverride Sub AddMenuRight(ByVal name As String, ByVal url As String, ByVal selected As Boolean, ByVal tooltip As String)
    Public MustOverride Sub AddMenuSide(ByVal name As String, ByVal url As String, ByVal selected As Boolean, ByVal tooltip As String)
    Public MustOverride Sub AddLinkSide(ByVal name As String, ByVal url As String, ByVal selected As Boolean, ByVal tooltip As String)

    Public MustOverride Property MenuTitle() As String
    Public MustOverride Property MenuDataSourceId() As String
    Public MustOverride Property MenuSideDataSourceID() As String
    Public MustOverride Property MenuSelected() As String
    Public Sub UnBindMenu()
        MenuDataSourceId = String.Empty
        MenuSideDataSourceID = String.Empty
    End Sub
    Public Sub UnbindSideMenu()
        MenuSideDataSourceID = String.Empty
    End Sub

    Public MustOverride Sub AddButton(ByVal url As String, ByVal tooltip As String)
End Class

File: WebsiteCSharp/App_Code/abstract/pages/CMasterPage.cs
using Microsoft.VisualBasic;
using System.Web.UI;
using System.Web.UI.HtmlControls;
using System.Web.UI.WebControls;

public abstract class CMasterPage : MasterPage
{

    protected void Page_PreRender(object sender, System.EventArgs e)
    {
        if (string.IsNullOrEmpty(PageHeading))
        {
            PageHeading = Page.Title;
        }

        if (string.IsNullOrEmpty(MenuTitle))
        {
            MenuTitle = CConfig.MenuTitle();
        }

        if (((Main.Controls.Count == 0) && (FieldSet != null)))
        {
            FieldSet.Visible = false;
        }

        if ((!Page.IsPostBack
                    && CSession.IsLoggedIn))
        {
            SchemaMembership.CClick.Log();
        }

    }

    public abstract ContentPlaceHolder Main { get; }
    public abstract HtmlGenericControl FieldSet { get; }


    public abstract string PageHeading { get; set; }
    public abstract void AddMenuLeft(string name, string url, bool selected, string tooltip );
    public abstract void AddMenuRight(string name, string url, bool selected, string tooltip);
    public abstract void AddMenuSide(string name, string url, bool selected, string tooltip);
    public abstract void AddLinkSide(string name, string url, bool selected, string tooltip);

    public abstract string MenuTitle{ get; set; }
    public abstract string MenuDataSourceId{ get; set; }
    public abstract string MenuSideDataSourceID{ get; set; }
    public abstract string MenuSelected{ get; set; }
    public void UnBindMenu()
    {
        MenuDataSourceId = string.Empty;
        MenuSideDataSourceID = string.Empty;

    }

    public void UnbindSideMenu()
    {
        MenuSideDataSourceID = string.Empty;
    }

    public abstract void AddButton(string url, string tooltip);
}